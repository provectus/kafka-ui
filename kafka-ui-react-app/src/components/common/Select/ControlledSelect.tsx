import * as React from 'react';
import { Controller } from 'react-hook-form';
import { FormError } from 'components/common/Input/Input.styled';
import { InputLabel } from 'components/common/Input/InputLabel.styled';
import { ErrorMessage } from '@hookform/error-message';

import Select, { SelectOption } from './Select';

interface ControlledSelectProps {
  name: string;
  label: React.ReactNode;
  hint?: string;
  options: SelectOption[];
  onChange?: (val: string | number) => void;
  disabled?: boolean;
  placeholder?: string;
}

const ControlledSelect: React.FC<ControlledSelectProps> = ({
  name,
  label,
  onChange,
  options,
  disabled = false,
  placeholder,
}) => {
  const id = React.useId();

  return (
    <div>
      <InputLabel htmlFor={id}>{label}</InputLabel>
      <Controller
        name={name}
        render={({ field }) => {
          return (
            <Select
              id={id}
              name={field.name}
              minWidth="270px"
              onChange={(value) => {
                if (onChange) onChange(value);
                field.onChange(value);
              }}
              value={field.value}
              options={options}
              placeholder={placeholder}
              disabled={disabled}
              ref={field.ref}
            />
          );
        }}
      />
      <FormError>
        <ErrorMessage name={name} />
      </FormError>
    </div>
  );
};

export default ControlledSelect;
