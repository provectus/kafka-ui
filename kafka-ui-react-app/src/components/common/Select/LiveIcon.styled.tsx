import styled, { useTheme } from 'styled-components';
import React from 'react';

interface Props {
  className?: string;
}

const SVGWrapper = styled.i`
  display: flex;
`;

const LiveIcon: React.FC<Props> = () => {
  const theme = useTheme();
  return (
    <SVGWrapper>
      <svg
        width="16"
        height="16"
        viewBox="0 0 16 16"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle cx="8" cy="8" r="7" fill={theme.icons.liveIcon.circleBig} />
        <circle cx="8" cy="8" r="4" fill={theme.icons.liveIcon.circleSmall} />
      </svg>
    </SVGWrapper>
  );
};

export default styled(LiveIcon)`
  position: absolute;
  left: 12px;
  top: 50%;
  transform: translateY(-50%);
  line-height: 0;
`;
