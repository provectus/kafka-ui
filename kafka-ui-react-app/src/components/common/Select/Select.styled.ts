import styled from 'styled-components';

interface Props {
  selectSize: 'M' | 'L';
  isLive?: boolean;
  minWidth?: string;
  disabled?: boolean;
  isThemeMode?: boolean;
}

interface OptionProps {
  disabled?: boolean;
}

export const Select = styled.ul<Props>`
  position: relative;
  list-style: none;
  display: flex;
  gap: 6px;
  align-items: center;
  justify-content: space-between;
  height: ${(props) => (props.selectSize === 'M' ? '32px' : '40px')};
  border: 1px
    ${({ theme, disabled, isThemeMode }) => {
      if (isThemeMode) {
        return 'none';
      }
      if (disabled) {
        return theme.select.borderColor.disabled;
      }

      return theme.select.borderColor.normal;
    }}
    solid;
  border-radius: 4px;
  font-size: 14px;
  width: fit-content;
  padding-left: 16px;
  padding-right: 12px;
  color: ${({ theme, disabled }) =>
    disabled ? theme.select.color.disabled : theme.select.color.normal};
  min-width: ${({ minWidth }) => minWidth || 'auto'};
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
  &:hover {
    color: ${({ theme, disabled }) =>
      disabled ? theme.select.color.disabled : theme.select.color.hover};
    border-color: ${({ theme, disabled }) =>
      disabled
        ? theme.select.borderColor.disabled
        : theme.select.borderColor.hover};
  }
  &:focus {
    outline: none;
    color: ${({ theme }) => theme.select.color.active};
    border-color: ${({ theme }) => theme.select.borderColor.active};
  }
  &:disabled {
    color: ${({ theme }) => theme.select.color.disabled};
    border-color: ${({ theme }) => theme.select.borderColor.disabled};
  }
`;

export const SelectedOptionWrapper = styled.div`
  display: flex;
  align-items: center;
  gap: 5px;
  overflow: hidden;
  text-overflow: ellipsis;
`;

export const OptionList = styled.ul`
  position: absolute;
  top: 100%;
  left: 0;
  max-height: 228px;
  margin-top: 4px;
  background-color: ${({ theme }) => theme.select.backgroundColor.normal};
  border: 1px ${({ theme }) => theme.select.borderColor.normal} solid;
  border-radius: 4px;
  font-size: 14px;
  line-height: 18px;
  color: ${({ theme }) => theme.select.color.normal};
  overflow-y: auto;
  z-index: 10;
  max-width: 300px;
  min-width: 100%;
  align-items: center;
  & div {
    white-space: nowrap;
  }
  &::-webkit-scrollbar {
    -webkit-appearance: none;
    width: 7px;
  }

  &::-webkit-scrollbar-thumb {
    border-radius: 4px;
    background-color: ${({ theme }) =>
      theme.select.optionList.scrollbar.backgroundColor};
  }

  &::-webkit-scrollbar:horizontal {
    height: 7px;
  }
`;

export const Option = styled.li<OptionProps>`
  display: flex;
  align-items: center;
  list-style: none;
  padding: 10px 12px;
  transition: all 0.2s ease-in-out;
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
  gap: 5px;
  color: ${({ theme, disabled }) =>
    theme.select.color[disabled ? 'disabled' : 'normal']};

  &:hover {
    background-color: ${({ theme, disabled }) =>
      theme.select.backgroundColor[disabled ? 'normal' : 'hover']};
  }

  &:active {
    background-color: ${({ theme }) => theme.select.backgroundColor.active};
  }
`;

export const SelectedOption = styled.li<{ isThemeMode?: boolean }>`
  display: flex;
  padding-right: ${({ isThemeMode }) => (isThemeMode ? '' : '16px')};
  list-style-position: inside;
  white-space: nowrap;
  & svg {
    path {
      fill: ${({ theme }) => theme.defaultIconColor};
    }
  }
  & div {
    display: none;
  }
`;
