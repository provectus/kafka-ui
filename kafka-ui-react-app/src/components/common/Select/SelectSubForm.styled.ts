import styled from 'styled-components';
import DatePicker from 'react-datepicker';

export const DatePickerInput = styled(DatePicker)`
  height: 32px;
  border: 1px ${({ theme }) => theme.select.borderColor.normal} solid;
  border-left: none;
  border-radius: 0 4px 4px 0;
  font-size: 14px;
  width: 100%;
  padding-left: 12px;
  background-color: ${({ theme }) => theme.input.backgroundColor.normal};
  color: ${({ theme }) => theme.input.color.normal};
  &::placeholder {
    color: ${({ theme }) => theme.input.color.normal};
  }

  background-image: url('data:image/svg+xml,%3Csvg width="10" height="6" viewBox="0 0 10 6" fill="none" xmlns="http://www.w3.org/2000/svg"%3E%3Cpath d="M1 1L5 5L9 1" stroke="%23454F54"/%3E%3C/svg%3E%0A') !important;
  background-repeat: no-repeat !important;
  background-position-x: 96% !important;
  background-position-y: 55% !important;
  appearance: none !important;

  &::react-datepicker {
    background-color: green;
  }
`;
