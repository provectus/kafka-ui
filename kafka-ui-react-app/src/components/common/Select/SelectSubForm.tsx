import * as React from 'react';
import { ReactDatePickerProps } from 'react-datepicker';
import Input, { InputProps } from 'components/common/Input/Input';
import DatePicker from 'components/common/DatePicker/DatePicker';

export interface SelectSubFormProps {
  inputType: typeof Input | typeof DatePicker;
  inputProps?: InputProps | Omit<ReactDatePickerProps, 'onChange'>;
  value?: string | Date | null;
  onChange?(value: string | Date | null): void;
}

const SelectSubForm: React.FC<SelectSubFormProps> = ({
  inputType,
  inputProps,
  value,
  onChange,
}) => {
  if (typeof Input === typeof inputType)
    return (
      <Input
        {...(inputProps as InputProps)}
        value={value as string}
        onChange={({ target: { value: val } }) => onChange && onChange(val)}
      />
    );
  if (typeof DatePicker === typeof inputType)
    return (
      <DatePicker
        {...(inputProps as ReactDatePickerProps)}
        selected={value as Date}
        onChange={(date) => onChange && onChange(date as Date)}
      />
    );
  return null;
};

export default SelectSubForm;
