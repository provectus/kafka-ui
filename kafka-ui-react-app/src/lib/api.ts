import {
  KsqlApi,
  TopicsApi,
  SchemasApi,
  BrokersApi,
  MessagesApi,
  ClustersApi,
  Configuration,
  KafkaConnectApi,
  ConsumerGroupsApi,
  AuthorizationApi,
  ApplicationConfigApi,
} from 'generated-sources';
import { BASE_PARAMS } from 'lib/constants';

const apiClientConf = new Configuration(BASE_PARAMS);

export const ksqlDbApiClient = new KsqlApi(apiClientConf);
export const topicsApiClient = new TopicsApi(apiClientConf);
export const brokersApiClient = new BrokersApi(apiClientConf);
export const schemasApiClient = new SchemasApi(apiClientConf);
export const messagesApiClient = new MessagesApi(apiClientConf);
export const clustersApiClient = new ClustersApi(apiClientConf);
export const kafkaConnectApiClient = new KafkaConnectApi(apiClientConf);
export const consumerGroupsApiClient = new ConsumerGroupsApi(apiClientConf);
export const authApiClient = new AuthorizationApi(apiClientConf);
export const appConfigApiClient = new ApplicationConfigApi(apiClientConf);
