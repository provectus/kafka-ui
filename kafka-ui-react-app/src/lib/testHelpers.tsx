import React, { PropsWithChildren, ReactElement, useMemo } from 'react';
import {
  MemoryRouter,
  MemoryRouterProps,
  Route,
  Routes,
} from 'react-router-dom';
import fetchMock from 'fetch-mock';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import { theme } from 'theme/theme';
import {
  render,
  renderHook,
  RenderOptions,
  waitFor,
} from '@testing-library/react';
import { AnyAction, Store } from 'redux';
import { RootState } from 'redux/interfaces';
import { configureStore } from '@reduxjs/toolkit';
import rootReducer from 'redux/reducers';
import {
  QueryClient,
  QueryClientProvider,
  UseQueryResult,
} from '@tanstack/react-query';
import { ConfirmContextProvider } from 'components/contexts/ConfirmContext';
import ConfirmationModal from 'components/common/ConfirmationModal/ConfirmationModal';
import { GlobalSettingsContext } from 'components/contexts/GlobalSettingsContext';
import { UserInfoRolesAccessContext } from 'components/contexts/UserInfoRolesAccessContext';

import { RolesType, modifyRolesData } from './permissions';

interface CustomRenderOptions extends Omit<RenderOptions, 'wrapper'> {
  preloadedState?: Partial<RootState>;
  store?: Store<Partial<RootState>, AnyAction>;
  initialEntries?: MemoryRouterProps['initialEntries'];
  userInfo?: {
    roles?: RolesType;
    rbacFlag: boolean;
  };
}

interface WithRouteProps {
  children: React.ReactNode;
  path: string;
}

export const expectQueryWorks = async (
  mock: fetchMock.FetchMockStatic,
  result: { current: UseQueryResult<unknown, unknown> }
) => {
  await waitFor(() => expect(result.current.isFetched).toBeTruthy());
  expect(mock.calls()).toHaveLength(1);
  expect(result.current.data).toBeDefined();
};

export const WithRoute: React.FC<WithRouteProps> = ({ children, path }) => {
  return (
    <Routes>
      <Route path={path} element={children} />
    </Routes>
  );
};

export const TestQueryClientProvider: React.FC<PropsWithChildren<unknown>> = ({
  children,
}) => {
  // use new QueryClient instance for each test run to avoid issues with cache
  const queryClient = new QueryClient({
    defaultOptions: { queries: { retry: false } },
  });
  return (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );
};

/**
 * @description it will create a UserInfo Provider that will actually
 * disable the rbacFlag , to user if you can pass it as an argument
 * */
const TestUserInfoProvider: React.FC<
  PropsWithChildren<{ data?: { roles?: RolesType; rbacFlag: boolean } }>
> = ({ children, data }) => {
  const contextValue = useMemo(() => {
    const roles = modifyRolesData(data?.roles);

    return {
      username: 'test',
      rbacFlag: !!(typeof data?.rbacFlag === 'undefined'
        ? false
        : data?.rbacFlag),
      roles,
    };
  }, [data]);

  return (
    <UserInfoRolesAccessContext.Provider value={contextValue}>
      {children}
    </UserInfoRolesAccessContext.Provider>
  );
};

const customRender = (
  ui: ReactElement,
  {
    preloadedState,
    store = configureStore<RootState>({
      reducer: rootReducer,
      preloadedState,
    }),
    initialEntries,
    userInfo,
    ...renderOptions
  }: CustomRenderOptions = {}
) => {
  // overrides @testing-library/react render.
  const AllTheProviders: React.FC<PropsWithChildren<unknown>> = ({
    children,
  }) => (
    <TestQueryClientProvider>
      <GlobalSettingsContext.Provider value={{ hasDynamicConfig: false }}>
        <ThemeProvider theme={theme}>
          <TestUserInfoProvider data={userInfo}>
            <ConfirmContextProvider>
              <Provider store={store}>
                <MemoryRouter initialEntries={initialEntries}>
                  <div>
                    {children}
                    <ConfirmationModal />
                  </div>
                </MemoryRouter>
              </Provider>
            </ConfirmContextProvider>
          </TestUserInfoProvider>
        </ThemeProvider>
      </GlobalSettingsContext.Provider>
    </TestQueryClientProvider>
  );
  return render(ui, { wrapper: AllTheProviders, ...renderOptions });
};

const customRenderHook = (hook: () => UseQueryResult<unknown, unknown>) =>
  renderHook(hook, { wrapper: TestQueryClientProvider });

export { customRender as render, customRenderHook as renderQueryHook };

export class EventSourceMock {
  url: string;

  close: () => void;

  open: () => void;

  error: () => void;

  onmessage: () => void;

  constructor(url: string) {
    this.url = url;
    this.open = jest.fn();
    this.error = jest.fn();
    this.onmessage = jest.fn();
    this.close = jest.fn();
  }
}
