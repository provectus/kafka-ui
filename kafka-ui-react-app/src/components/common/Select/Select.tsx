import React, { useState, useRef } from 'react';
import useClickOutside from 'lib/hooks/useClickOutside';
import DropdownArrowIcon from 'components/common/Icons/DropdownArrowIcon';

import * as S from './Select.styled';
import LiveIcon from './LiveIcon.styled';

export interface SelectProps {
  options?: Array<SelectOption>;
  id?: string;
  name?: string;
  selectSize?: 'M' | 'L';
  isLive?: boolean;
  minWidth?: string;
  value?: string | number;
  defaultValue?: string | number;
  placeholder?: string;
  disabled?: boolean;
  onChange?: (option: string | number) => void;
  isThemeMode?: boolean;
}

export interface SelectOption {
  label: string | number | React.ReactElement;
  value: string | number;
  disabled?: boolean;
  isLive?: boolean;
}

const Select = React.forwardRef<HTMLUListElement, SelectProps>(
  (
    {
      options = [],
      value,
      defaultValue,
      selectSize = 'L',
      placeholder = '',
      isLive,
      disabled = false,
      onChange,
      isThemeMode,
      ...props
    },
    ref
  ) => {
    const [selectedOption, setSelectedOption] = useState(value);
    const [showOptions, setShowOptions] = useState(false);

    const showOptionsHandler = () => {
      if (!disabled) setShowOptions(!showOptions);
    };

    const selectContainerRef = useRef(null);
    const clickOutsideHandler = () => setShowOptions(false);
    useClickOutside(selectContainerRef, clickOutsideHandler);

    const updateSelectedOption = (option: SelectOption) => {
      if (!option.disabled) {
        setSelectedOption(option.value);

        if (onChange) {
          onChange(option.value);
        }

        setShowOptions(false);
      }
    };

    React.useEffect(() => {
      setSelectedOption(value);
    }, [isLive, value]);

    return (
      <div ref={selectContainerRef}>
        <S.Select
          role="listbox"
          selectSize={selectSize}
          isLive={isLive}
          disabled={disabled}
          onClick={showOptionsHandler}
          onKeyDown={showOptionsHandler}
          isThemeMode={isThemeMode}
          ref={ref}
          tabIndex={0}
          {...props}
        >
          <S.SelectedOptionWrapper>
            {isLive && <LiveIcon />}
            <S.SelectedOption
              role="option"
              tabIndex={0}
              isThemeMode={isThemeMode}
            >
              {options.find(
                (option) => option.value === (defaultValue || selectedOption)
              )?.label || placeholder}
            </S.SelectedOption>
          </S.SelectedOptionWrapper>
          {showOptions && (
            <S.OptionList>
              {options?.map((option) => (
                <S.Option
                  value={option.value}
                  key={option.value}
                  disabled={option.disabled}
                  onClick={() => updateSelectedOption(option)}
                  tabIndex={0}
                  role="option"
                >
                  {option.isLive && <LiveIcon />}
                  {option.label}
                </S.Option>
              ))}
            </S.OptionList>
          )}
          <DropdownArrowIcon isOpen={showOptions} />
        </S.Select>
      </div>
    );
  }
);

Select.displayName = 'Select';

export default Select;
