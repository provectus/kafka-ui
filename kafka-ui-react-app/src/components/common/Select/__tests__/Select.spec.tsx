import Select, {
  SelectOption,
  SelectProps,
} from 'components/common/Select/Select';
import React from 'react';
import { render } from 'lib/testHelpers';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

jest.mock('react-hook-form', () => ({
  useFormContext: () => ({
    register: jest.fn(),
  }),
}));

const options: Array<SelectOption> = [
  { label: 'test-label1', value: 'test-value1', disabled: false },
  { label: 'test-label2', value: 'test-value2', disabled: true },
];

const renderComponent = (props?: Partial<SelectProps>) =>
  render(<Select name="test" {...props} />);

describe('Custom Select', () => {
  describe('when isLive is not specified', () => {
    beforeEach(() => {
      renderComponent({
        options,
      });
    });

    const getListbox = () => screen.getByRole('listbox');
    const getOption = () => screen.getByRole('option');

    it('renders component', () => {
      expect(getListbox()).toBeInTheDocument();
    });

    it('show select options when select is being clicked', async () => {
      expect(getOption()).toBeInTheDocument();
      await userEvent.click(getListbox());
      expect(screen.getAllByRole('option')).toHaveLength(3);
    });

    it('checking select option change', async () => {
      const optionLabel = 'test-label1';

      await userEvent.click(getListbox());
      await userEvent.selectOptions(getListbox(), [optionLabel]);

      expect(getOption()).toHaveTextContent(optionLabel);
    });

    it('trying to select disabled option does not trigger change', async () => {
      const normalOptionLabel = 'test-label1';
      const disabledOptionLabel = 'test-label2';

      await userEvent.click(getListbox());
      await userEvent.selectOptions(getListbox(), [normalOptionLabel]);
      await userEvent.click(getListbox());
      await userEvent.selectOptions(getListbox(), [disabledOptionLabel]);

      expect(getOption()).toHaveTextContent(normalOptionLabel);
    });
  });

  describe('when non-live', () => {
    it('there is not live icon', () => {
      renderComponent({ isLive: false });
      expect(screen.queryByTestId('liveIcon')).not.toBeInTheDocument();
    });
  });

  describe('when live', () => {
    it('there is live icon', () => {
      render(<Select name="test" {...{ isLive: true }} />);
      expect(screen.getByRole('listbox')).toBeInTheDocument();
    });
  });
});
