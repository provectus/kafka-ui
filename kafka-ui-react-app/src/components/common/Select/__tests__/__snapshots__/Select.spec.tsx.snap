// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Select when live matches the snapshot 1`] = `
<body>
  .c1 {
  position: absolute;
  left: 12px;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  line-height: 0;
}

.c2 {
  height: 40px;
  border: 1px #ABB5BA solid;
  border-radius: 4px;
  font-size: 14px;
  width: 100%;
  padding-left: 36px;
  padding-right: 16px;
  color: #171A1C;
  min-width: auto;
  background-image: url('data:image/svg+xml,%3Csvg width="10" height="6" viewBox="0 0 10 6" fill="none" xmlns="http://www.w3.org/2000/svg"%3E%3Cpath d="M1 1L5 5L9 1" stroke="%23454F54"/%3E%3C/svg%3E%0A') !important;
  background-repeat: no-repeat !important;
  background-position-x: calc(100% - 8px) !important;
  background-position-y: 55% !important;
  -webkit-appearance: none !important;
  -moz-appearance: none !important;
  appearance: none !important;
}

.c2:hover {
  color: #171A1C;
  border-color: #73848C;
}

.c2:focus {
  outline: none;
  color: #171A1C;
  border-color: #454F54;
}

.c2:disabled {
  color: #ABB5BA;
  border-color: #E3E6E8;
  cursor: not-allowed;
}

.c0 {
  position: relative;
}

<div>
    <div
      class="select-wrapper c0"
    >
      <i
        class="c1"
      >
        <svg
          fill="none"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="8"
            cy="8"
            fill="#FAD1D1"
            r="7"
          />
          <circle
            cx="8"
            cy="8"
            fill="#E51A1A"
            r="4"
          />
        </svg>
      </i>
      <select
        class="c2"
        role="listbox"
      />
    </div>
  </div>
</body>
`;

exports[`Custom Select when non-live matches the snapshot 1`] = `
.c1 {
  height: 40px;
  border: 1px #ABB5BA solid;
  border-radius: 4px;
  font-size: 14px;
  width: 100%;
  padding-left: 12px;
  padding-right: 16px;
  color: #171A1C;
  min-width: auto;
  background-image: url('data:image/svg+xml,%3Csvg width="10" height="6" viewBox="0 0 10 6" fill="none" xmlns="http://www.w3.org/2000/svg"%3E%3Cpath d="M1 1L5 5L9 1" stroke="%23454F54"/%3E%3C/svg%3E%0A') !important;
  background-repeat: no-repeat !important;
  background-position-x: calc(100% - 8px) !important;
  background-position-y: 55% !important;
  -webkit-appearance: none !important;
  -moz-appearance: none !important;
  appearance: none !important;
}

.c1:hover {
  color: #171A1C;
  border-color: #73848C;
}

.c1:focus {
  outline: none;
  color: #171A1C;
  border-color: #454F54;
}

.c1:disabled {
  color: #ABB5BA;
  border-color: #E3E6E8;
  cursor: not-allowed;
}

.c0 {
  position: relative;
}

<body>
  <div>
    <div
      class="select-wrapper c0"
    >
      <select
        class="c1"
        role="listbox"
      />
    </div>
  </div>
</body>
`;
